yellow-paper/docs/addresses-and-keys/addresses-and-keys.md

```diff
@@ -0,0 +1,903 @@
+
+The only alternative to this pattern is plume nullifiers, but there are tradeoffs, so we'll continue to provide support for non-plume nullifiers.
+
+> Note: not all nullifiers require injection of a secret _which is tied to a user's identity in some way_. Alternative examples of nullifier derivations (serving other use cases) are: hashing a storage slot; hashing a note hash and a secret contained within the note hash; ...?
+
+Requirements:
+- Support use cases where an app requires a secret "nullifier key" (linked to a user identity) to be derivable.
+  - Reason: it's a very common pattern.
+
+
+
+#### Is a nullifier key _pair_ needed?
+
+I.e. do we need both a nullifier secret key and a nullifier public key? Zcash sapling had both, but Zcash orchard (an upgrade) replaced the notion of a keypair with a single nullifier key. The [reason](https://zcash.github.io/orchard/design/keys.html) being:
+- _"[The nullifier secret key's (nsk's)] purpose in Sapling was as defense-in-depth, in case RedDSA [(the scheme used for signing txs, using the authentication secret key ask)] was found to have weaknesses; an adversary who could recover ask would not be able to spend funds. In practice it has not been feasible to manage nsk much more securely than a full viewing key [(dk, ak, nk, ovk)], as the computational power required to generate Sapling proofs has made it necessary to perform this step [(deriving nk from nsk)] on the same device that is creating the overall transaction (rather than on a more constrained device like a hardware wallet). We are also more confident in RedDSA now."_
+
+A nullifier public key might have the benefit (in Aztec) that a user could (optionally) provide their nullifier key nk to some 3rd party, to enable that 3rd party to see when the user's notes have been nullified for a particular app, without having the ability to nullify those notes.
+    - This presumes that within a circuit, the nk (not a public key; still secret!) would be derived from an nsk, and the nk would be injected into the nullifier.
+    - BUT, of course, it would be BAD if the nk were derivable as a bip32 normal child, because then everyone would be able to derive the nk from the master key, and be able to view whenever a note is nullified!
+    - The nk would need to ba a hardened key (derivable only from a secret).
+
+Given that it's acceptable to ZCash Orchard, we accept that a nullifier master secret key may be 'seen' by Aztec software.
+    
+
+### Diversified Addresses
+
+This is perhaps a "nice to have", although it's a core feature of zcash keys.
+
+Nice to have:
+- Alice can derive a diversified address; a random-looking address which she can (interactively) provide to Bob, so that Bob may send her funds (and general notes).
```

Comment:
Would this require a new account contract deployment? Or would it somehow "point" to the same account contract as the main address?

Reply:
It would still be able to point to the same account contract, although doing that 'linkage' as efficiently as possible, still needs thought.

There is still some thinking to be done for the most efficient way to derive stealth & diversified addresses. I can derive stealth & diversified public keys (elliptic curve points) which can be provably linked back to the master keys, and indeed to the main address, and to the main bytecode of the account contract. Lovely. But I haven't specified what a stealth/diversified _address_ looks like, yet.

A naive derivation would be the hash of the stealth/diversified incoming viewing public key. But that's a deviation from the way the _main_ address is derived. I considered making main address a hash of the master incoming viewing public key to achieve consistency of main and stealth/diversified address derivations, but it might not be an efficient way of structuring things.

An inefficient (but simple) derivation would be to hash all of the app-siloed public keys (matching the way the main address is currently derived). But the way the main address is currently derived isn't great.

I really meant what I said somewhere near the top of the spec: this works, but it's not optimised at all!

The main problem is: the number of constraints to link the bytecode to the main address is currently far fewer constraints than linking the bytecode to any app-siloed addresses (and even more constraints to link to stealth addresses). This 'asymmetry' (for want of a better word) is probably unavoidable, but it'll lead to circuits having to perform the number of constraints necessary to prove the 'longest path' (from master keys/bytecode, all the way to stealth address).

Reply:
I was actually meant to write something on this topic in the doc.

Reply:
> An inefficient (but simple) derivation would be to hash all of the app-siloed public keys (matching the way the main address is currently derived). But the way the main address is currently derived isn't great.

+1 from me on this one. I like having a single way of deriving an address as the hash of bytecode, salt, constructor args, and pubkeys; and then just change pubkeys from main to derived to get the derived address. This lets us treat all addresses the same.

Reply:
A few small fixes would be needed to derive the app-specific addresses in the same way, like the fact there are no app-siloed nullifier or outgoing viewing _public_ keys in this draft spec.